### ISSUE TITLE
[css-transforms] Named rotation axis

### ISSUE THREAD

**Issue by Crissov** [w3c/csswg-drafts#2130]
https://drafts.csswg.org/css-transforms-2/#individual-transforms

>     rotate: none | <number>{3}? <angle>

The default values for the numbers are `0 0 1` to get a planar 2D transform. I believe it would be cleaner and perhaps more readable if the axes had keywords: either `x`, `y` and `z` or `horizontal`, `vertical` and _?_.

I’m less sure whether `translate` should also support positional keywords like `top`, `right`, `bottom`, `left`, `center`, `start`, `end`. That would make this property even more similar to `position: relative` etc.

I understand that the three properties are named to match the functions in `transform` inherited. Consistency with those is probably more important than keeping with the general nomenclature used in CSS which prefers nouns over verbs for property names (i.e. `rotation` etc.).

**Comment by Nadya678** [comment]
Is it not intuitive the three numbers are x y z?

**Comment by AmeliaBR** [comment]
I agree that named axes would be preferable to the vector format for most simple cases. I would recommend using the x/y/z naming, because it is short and because it is consistent with the `rotateX`, `rotateY`, and `rotateZ` functions.

I think it would also be helpful to distinguish between a 3D rotation around the Z axis and a 2D rotation in the plane.  If you're already in 3D transformation mode, the effects are identical, but 2D versus 3D transformations have different effects in browser/GPU implementations, and in the parts of the spec that relate to SVG.

In other words, my ideal syntax would be something like:

```
rotate: none | [ 2d | x | y | z | <number>{3} ]? <angle> 
```

with `2d` as the default if an axis is not specified.

(Also, I'm not sure on how to write the grammar, but the order of angle and axis parameters should be interchangeable, assuming we aren't allowing unitless 0 for angles.)

**Comment by ewilligers** [comment]
To make the order interchangeable, use [`&&`](https://www.w3.org/TR/css-values-3/#component-combinators). Amelia's suggestion becomes

```
rotate: none | [ 2d | x | y | z | <number>{3} ]? && <angle> 
```

**Comment by Crissov** [comment]
`2d` would be parsed as a value with unknown unit, no? Perhaps `xy` instead, or `planar`.

**Comment by dbaron** [comment]
I think if we're not happy with the suggestions for the syntax of the `2d` vs `z` distinction that @AmeliaBR proposes, I think it's reasonable to not make that distinction -- and just say that if the author wants to force the transform to be 3-D rather than 2-D, they're going to have to do it some other way.  (And given the proposals so far, that would be my inclination, but I'm open to seeing something better.

I don't like `2d` since it's a dimension rather than a keyword (does `2.0d` or `2e0d` also work?), `xy` since it's the opposite of the other meanings (the plane rather than the axis perpendicular to it) and doesn't indicate at all which would be the 2-D or the 3-D version, and I also don't think `planar` is particularly clear about what the distinction between the values is.

Though another option is for the 2-D version to be what you get if you omit the axis/vector, and the 3-D version to be what you get if you include it.)

**Comment by AmeliaBR** [comment]
@Crissov Ugh, good point. `2d` is out.

I like the idea to make it that a specified axis means 3d mode, while the default (no axis specified) is 2d mode. 

If people want a keyword for 2d mode, `flat` might work. Or `two-d`. But from an authoring perspective, I think that if you're only working with 2d transformations, you'd probably want to leave it out altogether, and just do `rotate: 90deg`.

**Comment by tabatkins** [comment]
> Though another option is for the 2-D version to be what you get if you omit the axis/vector, and the 3-D version to be what you get if you include it.

That's what happens today, so yeah.  I do generally like the omitted value to default to some *specific* value (not be a distinct behavior in its own right), but if we can't come up with a good answer for what to call it, I'm fine with just leaving it out for now. We can always add a keyword for it later.

(I agree that `2d` is the best value, and also that it's impossible to use due to parsing concerns. Clearly the right answer is to use `d2` ^_^)

**Comment by Crissov** [comment]
Do people want to rotate around a 2D coordinate other than `0 0 1`? New issue?

**Comment by AmeliaBR** [comment]
@Crissov The rotation is always around the [`transform-origin`](https://drafts.csswg.org/css-transforms/#transform-origin-property) point.

**Comment by tabatkins** [comment]
Fixed in 09ed056
